libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,648 @@
+      std::get<_Idx>(__it_) -= static_cast<__underlying_diff_type>(__steps);
+    } else {
+      if (__offset >= __steps) {
+        std::get<_Idx>(__it_) -= static_cast<__underlying_diff_type>(__steps);
+      } else {
+        auto __prev_size = ranges::distance(std::get<_Idx - 1>(__parent_->__views_));
+        __it_.template emplace<_Idx - 1>(ranges::end(std::get<_Idx - 1>(__parent_->__views_)));
+        __advance_bwd<_Idx - 1>(__prev_size, __steps - __offset);
+      }
+    }
+  }
+
+  template <typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr auto __invoke_at_index(_Func&& __func) const {
+    // TODO(GCC 16): Just capture `this` when GCC PR113563 and PR121008 are fixed.
+    return [&__func, &__view_iter = *this]<std::size_t _Is>(this auto&& __self) {
+      if (_Is == __view_iter.__it_.index()) {
+        return __func.template operator()<_Is>();
+      }
+      if constexpr (_Is + 1 < sizeof...(_Views)) {
+        return __self.template operator()<_Is + 1>();
+      }
+      __builtin_unreachable();
+    }.template operator()<0>();
+  }
+
+  template <size_t... _Is, typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __apply_at_index(size_t __index, _Func&& __func, index_sequence<_Is...>) const {
+    ((__index == _Is ? (__func(integral_constant<size_t, _Is>{}), 0) : 0), ...);
+  }
+
+  template <size_t _Idx, typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr void __apply_at_index(size_t __index, _Func&& __func) const {
+    __apply_at_index(__index, std::forward<_Func>(__func), make_index_sequence<_Idx>{});
+  }
+
+  template <class... _Args>
+  _LIBCPP_HIDE_FROM_ABI explicit constexpr __iterator(__maybe_const<_Const, concat_view>* __parent, _Args&&... __args)
+    requires constructible_from<__base_iter, _Args&&...>
+      : __it_(std::forward<_Args>(__args)...), __parent_(__parent) {}
+
+  friend class concat_view;
+  friend class __iterator<!_Const>;
+
+public:
+  _LIBCPP_HIDE_FROM_ABI __iterator() = default;
+
+  _LIBCPP_HIDE_FROM_ABI constexpr __iterator(__iterator<!_Const> __i)
+    requires _Const && (convertible_to<iterator_t<_Views>, iterator_t<const _Views>> && ...)
+      : __it_([&__src = __i.__it_]<size_t... _Indices>(size_t __idx, index_sequence<_Indices...>) -> __base_iter {
+          _LIBCPP_ASSERT_VALID_ELEMENT_ACCESS(
+              !__src.valueless_by_exception(), "Trying to convert from a valueless iterator of concat_view.");
+          using __src_lref          = decltype((__src));
+          using __construction_fptr = __base_iter (*)(__src_lref);
+          static constexpr __construction_fptr __vtable[]{[](__src_lref __src_var) -> __base_iter {
+            return __base_iter(in_place_index<_Indices>, std::__unchecked_get<_Indices>(std::move(__src_var)));
+          }...};
+          return __vtable[__idx](__src);
+        }(__i.__it_.index(), make_index_sequence<variant_size_v<__base_iter>>{})),
+        __parent_(__i.__parent_) {}
+
+  _LIBCPP_HIDE_FROM_ABI constexpr decltype(auto) operator*() const {
```

Comment:
```suggestion
  [[nodiscard]] _LIBCPP_HIDE_FROM_ABI constexpr decltype(auto) operator*() const {
```

Looks like a candidate.